### PR TITLE
fix(mastra): relax output schema so API nulls and extra keys pass validation

### LINKED ISSUE #3047: Mastra provider: output schema validation rejects null values from third-party APIs
## Problem

`MastraProvider.wrapTool` passes `tool.outputParameters` directly to Mastra's `createTool` as `outputSchema`. Mastra validates tool output against this schema via `validateToolOutput`.

The problem: Composio's output schemas define strict non-nullable types (e.g., `{ type: "string" }`), but real third-party APIs (Linear, Notion, Jira, etc.) routinely return `null` for optional fields like `avatarUrl`, `description`, `assignee`, etc. When Mastra validates the output, it rejects the data and replaces it with an error — causing **data truncation**.

Additionally, `additionalProperties` defaults to `false` in the schema, so any unexpected fields also cause validation failures.

### Where it happens

In `ts/packages/providers/mastra/src/index.ts` (lines ~99-111):

```typescript
const outputSchema = applyCompatLayer({
  schema: tool.outputParameters ?? {},
  compatLayers: [],   // empty — no transformation
  mode: 'jsonSchema',
});

const mastraTool = createTool({
  id: tool.slug,
  description: tool.description ?? '',
  inputSchema,
  outputSchema,  // passed as-is, causes validation failures
  execute: async (inputData, _context) => { ... },
});
```

### Impact

Any Composio tool whose API response contains `null` for a field typed as a non-nullable primitive will have its output rejected by Mastra's validation. This affects a wide range of tools since most third-party APIs return `null` for optional/unset fields.

## Proposed solutions

Ranked by priority:

### 1. Quick fix — Omit `outputSchema` (trivial)

`outputSchema` is optional in Mastra's `createTool`. Simply stop passing it. For third-party API pass-through tools, output schema validation provides little value and causes real breakage.

```typescript
const mastraTool = createTool({
  id: tool.slug,
  description: tool.description ?? '',
  inputSchema,
  // outputSchema omitted — no validation
  execute: async (inputData, _context) => { ... },
});
```

### 2. Short-term — Add a compat layer for nullable handling (low effort)

The `applyCompatLayer` infrastructure already exists but is called with an empty `compatLayers` array. Create a compat layer (e.g., `makeOutputNullable`) that walks the schema and:
- Adds `"null"` to every primitive type union (`"string"` → `["string", "null"]`)
- Sets `additionalProperties: true` on object types

This keeps the architecture clean and lets other providers opt in too.

### 3. Long-term — Fix API-level output schemas (high effort)

The output schemas returned by the Composio API should accurately reflect what APIs actually return. If a field can be `null`, the schema should declare it as nullable. This is the "correct" fix but requires updating every tool's schema server-side and ongoing maintenance.

### 4. Nice-to-have — Provider config flag

Add an option like `outputSchemaMode: "strict" | "loose" | "none"` to provider options:
- `"none"` = omit outputSchema (option 1)
- `"loose"` = apply nullable compat layer (option 2)
- `"strict"` = current behavior

Gives users control without a breaking change.

## Related issues

- #2109
- #2716
- mastra-ai/mastra#11610

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by serhiizghama**
## Problem

Closes #3047.

`MastraProvider.wrapTool` passes each tool's `outputParameters` to Mastra's `createTool` as `outputSchema`, and Mastra validates every tool result against it via `validateToolOutput` — on a mismatch it **drops the data and substitutes an error**.

Composio's API-supplied output schemas are strict: optional fields are typed as non-nullable primitives (`{ type: "string" }`) and objects carry `additionalProperties: false`. Real third-party APIs (Linear, Notion, Jira, Slack, …) routinely return `null` for unset optional fields (`avatarUrl`, `assignee`, `description`, …) and occasionally extra keys. So strict validation rejects perfectly good responses and **truncates the tool output the model sees** — affecting a wide range of connector tools.

This is the recommended fix (#2) from the issue: output validation of opaque third-party pass-through payloads provides little value and causes real breakage.

## What this changes

A new `relaxOutputSchema` helper walks the (already dereferenced) output JSON schema and, non-mutating:

- makes every typed node nullable — `type: "string"` → `["string", "null"]` (handles string and array forms, idempotent)
- sets `additionalProperties: true` on object nodes that forbid (or omit) extra keys; a schema-valued `additionalProperties` is preserved and recursed into
- recurses through `properties`, `patternProperties`, `items`, `anyOf`/`oneOf`/`allOf`, `$defs`/`definitions`, etc.

`wrapTool` applies it to the output schema only (input schemas are untouched) before handing it to `@mastra/schema-compat`.

## Why this approach

Both transforms only **widen** what validates, so any payload that passed before still passes — there is no behavioural change for already-valid output, and no breaking change. It keeps the existing `dereferenceJsonSchema` → `applyCompatLayer` pipeline intact and adds one focused, unit-tested step. The `$ref`/`$defs` type-preservation behaviour from the existing regression suite is retained (the dereferenced inner type is kept, just made nullable).

## Testing

- New `test/relax-output-schema.test.ts` — 11 unit cases (nullable primitives, idempotency, nested objects, `additionalProperties` flip/preserve, array items, `anyOf` branches, empty schema, non-mutation, pass-through of non-objects).
- Updated existing `mastra.test.ts` / `mastra-ref.test.ts` assertions to reflect the relaxed output shape.
- `pnpm test` → 64/64 passing. `pnpm typecheck` and `pnpm build` clean; eslint/prettier clean on changed files.

### CODE REVIEW

File: ts/packages/providers/mastra/src/relax-output-schema.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  const originalType = result.type;
+  const isObjectSchema =
+    originalType === 'object' ||
+    (Array.isArray(originalType) && originalType.includes('object')) ||
+    'properties' in result ||
+    'patternProperties' in result;
+
+  // Make any typed node nullable so a `null` from the upstream API validates.
+  if ('type' in result) {
+    result.type = addNullToType(result.type);
```
Comment (sviat-barbutsa) [inline ts/packages/providers/mastra/src/relax-output-schema.ts:76]: Nice fix. I think this needs one more regression test for enum output fields becase if a schema is `{ type: "string", enum: ["open", "closed"] }`, this relaxation makes the type nullable, but `null` would still fail because it is not in the enum.

Could you, please, add a test for an enum-constrained output property returning `null`? The fix could either include `null` in the enum or use a nullable `anyOf` branch.

Comment (jkomyno) [reply ts/packages/providers/mastra/src/relax-output-schema.ts:76]: Good catch — fixed in 9a3153516. `relaxOutputSchema` now widens `enum`/`const` to also admit `null` (`{ type: "string", enum: ["open","closed"] }` → `enum: ["open","closed",null]`; a `const` becomes a two-member nullable `enum`), since `enum`/`const` constrain the value independently of `type`, so making the type nullable alone was not enough.

Covered by new tests:
- unit (`relax-output-schema.test.ts`): "widens an enum to also admit null…" and "converts const into a nullable two-member enum"
- integration (`output-validation.integration.test.ts`, real `@mastra/schema-compat`): a `null` enum value now validates, while a bogus value (`"nope"`) is still rejected — so the constraint is widened, not removed.

File: ts/packages/providers/mastra/src/relax-output-schema.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  }
```
[the remark this answers is not in the document]
Comment (jkomyno) [reply ts/packages/providers/mastra/src/relax-output-schema.ts:77]: Fixed in 9a3153516. `relaxOutputSchema` now widens `enum`/`const` to also admit `null` — `enum` gets `null` appended (idempotently), and a `const` is converted to a two-member nullable `enum` (`[const, null]`). This closes the gap where a nullable `type` alone still rejected a `null` because the `enum`/`const` kept constraining the value.

Verified end-to-end against the real `@mastra/schema-compat` validation path (`relaxOutputSchema → applyCompatLayer → convertSchemaToZod → parse`): a `null` enum value now passes, a bogus value still fails.

File: ts/packages/providers/mastra/src/relax-output-schema.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    result.additionalProperties = relaxOutputSchema(result.additionalProperties);
+  } else if (
+    isObjectSchema &&
+    (result.additionalProperties === false || result.additionalProperties === undefined)
+  ) {
+    result.additionalProperties = true;
+  }
+
+  for (const key of SUBSCHEMA_KEYS) {
+    if (key in result) result[key] = relaxOutputSchema(result[key]);
```
[the remark this answers is not in the document]
Comment (jkomyno) [reply ts/packages/providers/mastra/src/relax-output-schema.ts:107]: Fixed in 9b71ae589. `not` is now excluded from the recursed sub-schema keys, so its subschema is preserved verbatim. Since `not` negates its subschema, relaxing the inner schema (e.g. making it nullable) would *narrow* what the parent admits — a value valid under `not: { type: "string" }` (like `null`) would start failing — which is the opposite of this helper's only-widen guarantee.

Verified by a regression test (both unit and integration): under `not: { type: "string" }`, `null` and `42` stay valid and `"str"` is still rejected, before and after relaxation.

**Review by jkomyno — APPROVED** [review]

**Comment by jkomyno** [comment]
Thank you! I pushed a couple of fixes, so we're ready to merge
